decryptor/postgresql/prepared_statements_sql_observer.go

```diff
@@ -0,0 +1,172 @@
+	logrus.Debugln("PreparedStatementsQuery.Prepare")
+
+	var preparedStatementName = prepareQuery.PreparedStatementName.String()
+	var registry = encryptor.session.PreparedStatementRegistry()
+
+	// PostgreSQL allows create statement only once during the session
+	if _, err := registry.StatementByName(prepareQuery.PreparedStatementName.String()); err == nil {
+		logrus.WithField(logging.FieldKeyEventCode, logging.EventCodeErrorGeneral).
+			WithError(err).Errorln("PreparedStatement already stored in registry")
+		return nil, false, ErrStatementAlreadyInRegistry
+	}
+
+	var stmt = prepareQuery.PreparedStatementQuery.(sqlparser.Statement)
+	var preparedStatement = NewPreparedStatement(preparedStatementName, sqlparser.String(stmt), stmt)
+
+	if err := registry.AddStatement(preparedStatement); err != nil {
+		logrus.WithField(logging.FieldKeyEventCode, logging.EventCodeErrorGeneral).
+			WithError(err).Errorln("Failed to add prepared statement")
+		return nil, false, err
+	}
+
+	// run OnQuery with inner query of prepared statement
+	changedObject, changed, err := encryptor.queryObserver.OnQuery(ctx, base.NewOnQueryObjectFromStatement(stmt, encryptor.parser))
+	if err != nil {
+		return nil, false, err
+	}
+
+	// if the inner query of prepared statement was changed, replace it in sqlparser.PreparedQuery
+	if changed {
+		changedStatement, err := changedObject.Statement()
+		if err != nil {
+			return nil, false, err
+		}
+
+		changedPreparedQuery, ok := changedStatement.(sqlparser.PreparedQuery)
+		if !ok {
+			logrus.WithField(logging.FieldKeyEventCode, logging.EventCodeErrorGeneral).
+				WithError(err).Errorln("Failed to cast changedStatement to PreparedQuery")
+			return nil, false, err
+		}
+		prepareQuery.PreparedStatementQuery = changedPreparedQuery
+	}
+
+	logrus.WithField("prepared_name", preparedStatementName).Debug("Registered new prepared statement")
+	return base.NewOnQueryObjectFromStatement(prepareQuery, encryptor.parser), changed, nil
+}
+
+func (encryptor *PreparedStatementsQuery) onExecute(ctx context.Context, executeQuery *sqlparser.Execute) (base.OnQueryObject, bool, error) {
+	logrus.Debugln("PreparedStatementsQuery.Execute")
+
+	var preparedStatementName = executeQuery.PreparedStatementName.String()
+	var registry = encryptor.session.PreparedStatementRegistry()
+
+	preparedStatement, err := registry.StatementByName(preparedStatementName)
+	if err != nil {
+		logrus.WithField(logging.FieldKeyEventCode, logging.EventCodeErrorGeneral).
+			WithError(err).Errorln("PreparedStatement not present in registry")
+		return nil, false, ErrStatementNotPresentInRegistry
+	}
+
+	// collect all BoundValues
+	values := make([]base.BoundValue, 0, len(executeQuery.Values))
+	for _, value := range executeQuery.Values {
+		switch val := value.(type) {
+		case *sqlparser.SQLVal:
+			values = append(values, NewPgBoundValue(val.Val, bindFormatText))
+		case *sqlparser.NullVal:
```

Comment:
should we handle `default` case with other types and append `nil` values to be able to map them to appropriate placeholders in the query?

Reply:
below I see that we collects placeholders by their indexes and only interesting for us. but we should be able to fill array of values with not interesting to keep correct indexing.